src/node_contextify.cc

```diff
@@ -441,40 +373,111 @@ class ContextifyContext {
     bool is_function = value->IsFunction();
 
     if (!is_declared && args.ShouldThrowOnError() && is_contextual_store &&
-    !is_function)
+        !is_function)
       return;
 
     ctx->sandbox()->Set(property, value);
   }
 
 
-  static void GlobalPropertyQueryCallback(
+  static void GlobalPropertyDescriptorCallback(
       Local<Name> property,
-      const PropertyCallbackInfo<Integer>& args) {
+      const PropertyCallbackInfo<Value>& info) {
     ContextifyContext* ctx;
-    ASSIGN_OR_RETURN_UNWRAP(&ctx, args.Data().As<Object>());
+    ASSIGN_OR_RETURN_UNWRAP(&ctx, info.Data().As<Object>());
 
     // Still initializing
     if (ctx->context_.IsEmpty())
       return;
 
     Local<Context> context = ctx->context();
-    Maybe<PropertyAttribute> maybe_prop_attr =
-        ctx->sandbox()->GetRealNamedPropertyAttributes(context, property);
+    v8::Isolate* isolate = context->GetIsolate();
+    HandleScope scope(isolate);
 
-    if (maybe_prop_attr.IsNothing()) {
-      maybe_prop_attr =
-          ctx->global_proxy()->GetRealNamedPropertyAttributes(context,
-                                                              property);
-    }
+    Local<String> key = property->ToString(isolate);
+    Local<Object> sandbox = ctx->sandbox();
 
-    if (maybe_prop_attr.IsJust()) {
-      PropertyAttribute prop_attr = maybe_prop_attr.FromJust();
-      args.GetReturnValue().Set(prop_attr);
+    Maybe<bool> has = sandbox->HasOwnProperty(context, key);
+    Local<Value> descriptor_intercepted = has.IsNothing() ?
+        ctx->global_proxy()->GetOwnPropertyDescriptor(context, key)
+            .ToLocalChecked() :
+        sandbox->GetOwnPropertyDescriptor(context, key).ToLocalChecked();
```

Comment:
I'd write this as an if statement.  It's rather hard to read.

Reply:
Agree. Done.